plugin/metadata/metadata_test.go

```diff
@@ -0,0 +1,57 @@
+package metadata
+
+import (
+	"context"
+	"testing"
+
+	"github.com/coredns/coredns/plugin/test"
+	"github.com/miekg/dns"
+)
+
+// testMetadater implements fake Metadaters (plugins which inmplement Metadater interface)
+type testMetadater struct {
+	key   interface{}
```

Comment:
Each plugin can't define its own key - if we do that then there is no way for things that are going to *use* the metadata - like `rewrite` and `policy` to pull the data out. Instead, the `metadata` plugin should define a key, and methods that the plugins can use to add/remove their metadata from the context. I also suggest that internally this is just kept by `metadata` as a map in the context; similar to how [grpc metadata](https://github.com/grpc/grpc-go/blob/master/Documentation/grpc-metadata.md) works.